### PR TITLE
[SPARK-7012][SQL] Add support for NOT NULL modifier for column definitions on DDLParser

### PR DESCRIPTION
**Pull request by sabhyankar**
Add support for NOT NULL modifier for column definitions in DDLParser

### CODE REVIEW

**Comment by sabhyankar** [comment]
@smola I have created this PR for SPARK-7012

**Comment by smola** [comment]
@sabhyankar Great! The implementation looks good. Could you add a test case for it?

**Comment by sabhyankar** [comment]
@smola Thanks for the review. I will add a test case for it and push another commit today.

**Comment by sabhyankar** [comment]
@smola I have added a test case for this. Let me know if you see anything else that needs to be changed or added.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/DDLTestSuite.scala [hunk 1]
```diff
@@ -113,4 +124,20 @@ class DDLTestSuite extends DataSourceTest with SharedSQLContext {
     assert(attributes.map(_.name) === Seq("col_name", "data_type", "comment"))
     assert(attributes.map(_.dataType).toSet === Set(StringType))
   }
+
+  test("SPARK-7012 Create table statement should support NOT NULL modifier for columns") {
+    sql(
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/DDLTestSuite.scala]: you can use `withTempPath` for this test. [example](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala#L1732-L1765)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DDLParser.scala [hunk 2]
```diff
@@ -173,13 +174,15 @@ class DDLParser(parseQuery: String => LogicalPlan)
...
-    ident ~ dataType ~ (COMMENT ~> stringLit).?  ^^ { case columnName ~ typ ~ cm =>
-      val meta = cm match {
-        case Some(comment) =>
-          new MetadataBuilder().putString(COMMENT.str.toLowerCase, comment).build()
-        case None => Metadata.empty
-      }
-
-      StructField(columnName, typ, nullable = true, meta)
+    ident ~ dataType ~ (NOT ~ NULL).? ~ (COMMENT ~> stringLit).?  ^^ {
+      case columnName ~ typ ~ setNotNullable ~ cm =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DDLParser.scala]: nit: I feel `notNull` is better than `setNotNullable`.

Comment (sabhyankar) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DDLParser.scala]: I guess that makes it a little less ambiguous. I have made the change in the latest commit.

File: sql/core/src/test/scala/org/apache/spark/sql/sources/DDLTestSuite.scala (hunk 1 above)
Comment (sabhyankar) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/DDLTestSuite.scala]: Thanks @cloud-fan ! This makes it much cleaner. I have made the change in the latest commit.

**Comment by sabhyankar** [comment]
@smola @cloud-fan Thanks again for the review guys. I have made the additional changes that were noted.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DDLParser.scala [hunk 3]
```diff
@@ -173,13 +174,15 @@ class DDLParser(parseQuery: String => LogicalPlan)
...
-      StructField(columnName, typ, nullable = true, meta)
+    ident ~ dataType ~ (NOT ~ NULL).? ~ (COMMENT ~> stringLit).?  ^^ {
+      case columnName ~ typ ~ notNull ~ cm =>
+        val meta = cm match {
+          case Some(comment) =>
+            new MetadataBuilder().putString(COMMENT.str.toLowerCase, comment).build()
+          case None => Metadata.empty
+        }
+
+        val isNullable = !notNull.isDefined
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DDLParser.scala]: nit: `!xxx.isDefined` == `xxx.isEmpty`

File: sql/core/src/test/scala/org/apache/spark/sql/sources/DDLTestSuite.scala [hunk 4]
```diff
@@ -17,6 +17,7 @@
 
 package org.apache.spark.sql.sources
 
+
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/DDLTestSuite.scala]: nit: remove this extra blank line.

**Comment by cloud-fan** [comment]
LGTM

**Comment by sabhyankar** [comment]
Thanks again @cloud-fan I have pushed a new commit with the updates!

**Comment by sabhyankar** [comment]
Hi @smola - Is there anything else that you need me to check for this PR?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/DDLTestSuite.scala [hunk 5]
```diff
@@ -113,4 +113,23 @@ class DDLTestSuite extends DataSourceTest with SharedSQLContext {
...
+         |( tCol1 INT NOT NULL,
+         |  tCol2 STRING
+         |)
+         |USING parquet
+         |OPTIONS (
+         |  path '$path'
+         |)
+       """.stripMargin
+      )
+      caseInsensitiveContext.dropTempTable("tempTableDDL")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/DDLTestSuite.scala]: improve this test case to make sure we really make that column not nullable, i.e. read the table out and check its schema.

Comment (smola) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/DDLTestSuite.scala]: @sabhyankar Do you have time to modify this? I think that once the test is improved with @cloud-fan comment, the change is probably good to merge.

Comment (sabhyankar) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/DDLTestSuite.scala]: Hi @smola @cloud-fan - I will take a look at the test case this week and update you guys

**Comment by yhuai** [comment]
I am not sure if we should do it now because it is still possible to have nulls in a column defined as `NOT NULL`.

**Comment by yhuai** [comment]
How about we close this PR for now and revisit it later?

**Pull request closed without merging**